File: ruby/lazyk_test.rb
require "minitest/autorun"
require "stringio"
require_relative "./lazyk.rb"

class TestLazyK < Minitest::Test

  def test_parse
    assert_equal "S", LazyK::Parser.parse("s").to_s
    assert_equal "K", LazyK::Parser.parse("k").to_s
    assert_equal "I", LazyK::Parser.parse("i").to_s
    assert_equal "((S K) K)", LazyK::Parser.parse("SKK").to_s
    assert_equal "(S (K K))", LazyK::Parser.parse("S(KK)").to_s
    assert_equal "((S K) K)", LazyK::Parser.parse("``skk").to_s
    assert_equal "(S (K K))", LazyK::Parser.parse("`s`kk").to_s
    assert_equal "(Iota ((Iota Iota) Iota))", LazyK::Parser.parse("*i**iii").to_s
    assert_equal "((I (Iota I)) I)", LazyK::Parser.parse("`i*iIi").to_s
    assert_equal "(S (K ((I S) K)))", LazyK::Parser.parse("01").to_s
    assert_equal "(S (((S K) ((S (K ((I S) K))) Iota)) I))", LazyK::Parser.parse("`s(SK*01ii)").to_s
    assert_equal "(S (K K))", LazyK::Parser.parse(" S # l1\n  (K # l2\n\n K # l4\n ) # l5").to_s
    assert_equal "I", LazyK::Parser.parse("").to_s
    assert_equal "((K I) I)", LazyK::Parser.parse("K()()").to_s

    assert_raises(SyntaxError) { LazyK::Parser.parse("SNK") }
    assert_raises(SyntaxError) { LazyK::Parser.parse("(SK") }
    assert_raises(SyntaxError) { LazyK::Parser.parse("``sk") }
    assert_raises(SyntaxError) { LazyK::Parser.parse("*i") }
  end

  def test_comb
    s = LazyK::Expr::S
    k = LazyK::Expr::K
    i = LazyK::Expr::I
    iota = LazyK::Expr::IOTA
    f = LazyK::Expr::FALSE
    n1 = LazyK::Expr.new(:num, 1)
    n2 = LazyK::Expr.new(:num, 2)
    assert_equal n1, i.apply(n1).eval
    assert_equal n1, k.apply(n1).apply(n2).eval
    assert_equal n1, s.apply(k).apply(k).apply(n1).eval
    assert_equal n2, f.apply(n1).apply(n2).eval
    assert_equal s, iota.apply(k).eval
    assert_equal k, iota.apply(f).eval
  end

  def test_primitive_cnum
    [0, 1, 2, 128, 256].each{|i|
      assert_equal i, LazyK::Expr::CNUMS[i].to_i
    }
    cn2 = LazyK::Expr::CNUMS[2]
    cn3 = LazyK::Expr::CNUMS[3]
    assert_equal 8, cn3.apply(cn2).to_i
    assert_equal 64, cn2.apply(cn3.apply(cn2)).to_i
    assert_equal 256, cn3.apply(cn2).apply(cn2).to_i
  end

  def test_composed_cnum
    assert_equal 0, LazyK::Expr::FALSE.to_i
    assert_equal 1, LazyK::Expr::I.to_i
    succ = LazyK::Parser.parse("S(S(KS)K)")
    cn2 = succ.apply(LazyK::Expr::I)
    cn3 = succ.apply(cn2)
    assert_equal 2, cn2.to_i
    assert_equal 3, cn3.to_i
    assert_equal 8, cn3.apply(cn2).to_i
    assert_equal 64, cn2.apply(cn3.apply(cn2)).to_i
    assert_equal 256, cn3.apply(cn2).apply(cn2).to_i
  end

  def test_input
    stdin = StringIO.new("bar\n")
    stdout = StringIO.new
    car = ->(e){ e.apply(LazyK::Expr::TRUE) }
    cdr = ->(e){ e.apply(LazyK::Expr::FALSE) }
    m = LazyK.new(stdin, stdout)
    input = LazyK::Expr.new(:read, m.method(:readc))
    assert_equal 98, car[input].to_i
    assert_equal 97, car[cdr[input]].to_i
    assert_equal 114, car[cdr[cdr[input]]].to_i
    assert_equal 10, car[cdr[cdr[cdr[input]]]].to_i
    assert_equal 256, car[cdr[cdr[cdr[cdr[input]]]]].to_i
  end

  def test_output
    cons = LazyK::Parser.parse("S(S(KS)(S(KK)(S(KS)(S(K(SI))K))))(KK)")
    expect = "A\nBC"
    l = (expect.bytes + [256, 0]).reverse.map{|i| LazyK::Expr::CNUMS[i] }.reduce{|a, c| cons.apply(c).apply(a) }
    stdin = StringIO.new("")
    stdout = StringIO.new
    LazyK.new(stdin, stdout).send(:print_list, l)
    assert_equal expect, stdout.string
  end

  def run_lazyk(src, instr = "")
    stdin = StringIO.new(instr)
    stdout = StringIO.new
    result = LazyK.new(stdin, stdout).run(src)
    [result, stdout.string]
  end
  
  def test_program
    # echo
    assert_equal [0, "asdf\nqwer"], run_lazyk("``skk", "asdf\nqwer")

    # cdr
    assert_equal [0, "jkl"], run_lazyk("SI`k`ki", "hjkl")

    # (lambda (input) (cons (* 8 (car input)) k))
    assert_equal [8, ""], run_lazyk("11111110001111111000111100111110001111111000111100111111000111111111000001111111000111100111001111111000111100111111000111100111111110001111111000111100111110001110011111110001111111000111100111110001110011111111100000111111100011111110001111001111100011100111111111000001111111000111111111000001111001110011110011110011100", "!")
  end
end

File: python/lazyk_test.py
import unittest
from io import BytesIO
from functools import reduce
from lazyk import EType, Expr, Parser, LazyK


class Util:
    NK = LazyK(None, None)

    @staticmethod
    def cn2i(e: Expr) -> int:
        return Util.NK.cn_to_int(e)


class ParserTestCase(unittest.TestCase):
    def test_single_token(self):
        self.assertEqual(Parser.parse("s"), LazyK.S)
        self.assertEqual(Parser.parse("k"), LazyK.K)
        self.assertEqual(Parser.parse("i"), LazyK.I)

    def test_applications(self):
        self.assertEqual(str(Parser.parse("SKK")), "((S K) K)")
        self.assertEqual(str(Parser.parse("S(KK)")), "(S (K K))")
        self.assertEqual(str(Parser.parse("``skk")), "((S K) K)")
        self.assertEqual(str(Parser.parse("`s`kk")), "(S (K K))")
        self.assertEqual(str(Parser.parse("*i**iii")), "(Iota ((Iota Iota) Iota))")
        self.assertEqual(str(Parser.parse("`i*iIi")), "((I (Iota I)) I)")
        self.assertEqual(str(Parser.parse("01")), "(S (K ((I S) K)))")
        self.assertEqual(
            str(Parser.parse("`s(SK*01ii)")), "(S (((S K) ((S (K ((I S) K))) Iota)) I))"
        )

    def test_spaces_and_comments(self):
        self.assertEqual(
            str(Parser.parse(" S # l1\n  (K # l2\n\n K # l4\n ) # l5")), "(S (K K))"
        )

    def test_blank(self):
        self.assertEqual(Parser.parse(""), LazyK.I)
        self.assertEqual(str(Parser.parse("K()()")), "((K I) I)")

    def test_syntax_error(self):
        self.assertRaises(Exception, lambda: Parser.parse("SNK"))
        self.assertRaises(Exception, lambda: Parser.parse("(SK"))
        self.assertRaises(Exception, lambda: Parser.parse("``sk"))
        self.assertRaises(Exception, lambda: Parser.parse("*i"))


class CombinatorTestCase(unittest.TestCase):
    def combinator_test(self):
        s = LazyK.S
        k = LazyK.K
        i = LazyK.I
        iota = LazyK.IOTA
        f = LazyK.CFALSE
        t = LazyK.CTRUE
        n1 = LazyK.CNUMS[1]
        n2 = LazyK.CNUMS[2]
        self.assertEqual(LazyK.eval_expr(i.apply(n1)), n1)
        self.assertEqual(LazyK.eval_expr(k.apply(n1).apply(n2)), n1)
        self.assertEqual(LazyK.eval_expr(s.apply(n1).apply(k).apply(k).apply(n1)), n1)
        self.assertEqual(LazyK.eval_expr(f.apply(n1).apply(n2)), n2)
        self.assertEqual(LazyK.eval_expr(iota.apply(t)), s)
        self.assertEqual(LazyK.eval_expr(iota.apply(f)), k)


class ChurchNumTestCase(unittest.TestCase):
    def test_primitive_churchnum(self):
        for i in range(256):
            self.assertEqual(Util.cn2i(LazyK.CNUMS[i]), i)
        cn2 = LazyK.CNUMS[2]
        cn3 = LazyK.CNUMS[3]
        self.assertEqual(Util.cn2i(cn3.apply(cn2)), 8)
        self.assertEqual(Util.cn2i(cn2.apply(cn3.apply(cn2))), 64)
        self.assertEqual(Util.cn2i(cn3.apply(cn2).apply(cn2)), 256)

    def test_composed_churchnum(self):
        s = LazyK.S
        k = LazyK.K
        i = LazyK.I
        f = LazyK.CFALSE
        self.assertEqual(Util.cn2i(f), 0)
        self.assertEqual(Util.cn2i(i), 1)

        succ = s.apply(s.apply(k.apply(s)).apply(k))
        cn2 = succ.apply(i)
        cn3 = succ.apply(cn2)
        self.assertEqual(Util.cn2i(cn2), 2)
        self.assertEqual(Util.cn2i(cn3), 3)
        self.assertEqual(Util.cn2i(cn3.apply(cn2)), 8)
        self.assertEqual(Util.cn2i(cn2.apply(cn3.apply(cn2))), 64)
        self.assertEqual(Util.cn2i(cn3.apply(cn2).apply(cn2)), 256)


class IOTestCase(unittest.TestCase):
    def test_input(self):
        input = BytesIO(b"bar\n")
        output = BytesIO()
        m = LazyK(input, output)
        car = lambda e: e.apply(LazyK.CTRUE)
        cdr = lambda e: e.apply(LazyK.CFALSE)
        cn2i = lambda e: m.cn_to_int(e)
        r = Expr(EType.Read)
        self.assertEqual(cn2i(car(r)), 98)
        self.assertEqual(cn2i(car(cdr(r))), 97)
        self.assertEqual(cn2i(car(cdr(cdr(r)))), 114)
        self.assertEqual(cn2i(car(cdr(cdr(cdr(r))))), 10)
        self.assertEqual(cn2i(car(cdr(cdr(cdr(cdr(r)))))), 256)

    def test_output(self):
        cons = Parser.parse("S(S(KS)(S(KK)(S(KS)(S(K(SI))K))))(KK)")
        o = [97, 10, 98, 99, 256]
        list.reverse(o)
        l = reduce(lambda a, i: cons.apply(LazyK.CNUMS[i]).apply(a), o, LazyK.NUM0)
        input = BytesIO()
        output = BytesIO()
        m = LazyK(input, output)
        m.print_list(l)
        self.assertEqual(output.getvalue(), b"a\nbc")


class ProgramTestCases(unittest.TestCase):
    def run_lazyk(self, src, input):
        output = BytesIO()
        m = LazyK(BytesIO(input), output)
        result = m.run(src)
        return (result, output.getvalue())

    def test_program(self):
        self.assertEqual(self.run_lazyk("SKK", b"asdf\nqwer"), (0, b"asdf\nqwer"))
        self.assertEqual(self.run_lazyk("SI`k`ki", b"hjkl"), (0, b"jkl"))
        self.assertEqual(
            self.run_lazyk(
                "11111110001111111000111100111110001111111000111100111111000111111111000001111111000111100111001111111000111100111111000111100111111110001111111000111100111110001110011111110001111111000111100111110001110011111111100000111111100011111110001111001111100011100111111111000001111111000111111111000001111001110011110011110011100",
                b"!",
            ),
            (8, b""),
        )


if __name__ == "__main__":
    unittest.main(verbosity=2)
